### PR TITLE
refactor boto_vpc to cache ids and eliminate duplicate code

### PR DESCRIPTION
**Pull request by mgwilliams**
Due to the cascading issues with #22068, I'm replacing it with this more backwards-compatible PR.

Ping @ryan-lane and @claudiupopescu

### CODE REVIEW

**Comment by mgwilliams** [comment]
This last commit should fix the boto_vpc related test failure... 
@rallytime - not sure what these other failures are.

**Comment by rallytime** [comment]
@mgwilliams Thanks for cleaning that up. The other test failures are a result of a race condition in our test suite that we are working on tracking down, so no need to worry about those. :)

File: salt/modules/boto_vpc.py [hunk 1]
```diff
@@ -1527,8 +1859,51 @@ def describe_vpcs(vpc_id=None, name=None, cidr=None, tags=None,
...
 
-def describe_subnets(subnet_ids=None, vpc_id=None, cidr=None, region=None, key=None, keyid=None,
-                     profile=None):
+def describe_subnet(subnet_id=None, region=None, key=None, keyid=None, profile=None):
+    '''
+    Given a subnet id, describe it's properties.
+
+    Returns a dictionary of interesting properties.
+    CLI example::
+
```
Comment (iggy) [inline salt/modules/boto_vpc.py]: Version added?

File: salt/states/boto_vpc.py [hunk 2]
```diff
@@ -125,25 +125,82 @@ def present(name, cidr_block, instance_tenancy=None, dns_support=None,
     return ret
 
 
-def subnet_present(name, vpc_id, cidr_block, availability_zone=None, tags=None, region=None,
+def absent(name, tags=None, region=None, key=None, keyid=None, profile=None):
+    '''
```
Comment (iggy) [inline salt/states/boto_vpc.py]: subnet_absent or vpc_absent? And version...

**Pull request merged**